Manual/VCGen.lean

```diff
@@ -28,1004 +29,884 @@ set_option mvcgen.warning false
-  -- property we wanted
-  case vc3 h =>
-    grind
-```
-
-:::paragraph
-Note that the case labels are actually unique prefixes of the full case labels.
-Whenever referring to cases, only this prefix should be used; the suffix is merely a hint to the user of where that particular {tech}[VC] came from.
-For example:
-
-* `vc1.step` conveys that this {tech}[VC] proves the inductive step for the loop
-* `vc2.a.pre` is meant to prove that the hypotheses of a goal imply the precondition of a specification (of {name}`forIn`).
-* `vc3.a.post.success` is meant to prove that the postcondition of a specification (of {name}`forIn`) implies the desired property.
-:::
+{docstring Std.Do.SPred}
 
-:::paragraph
-After specifying the loop invariant, the proof can be shortened to just {keyword}`all_goals mleave; grind` (where {tactic}`mleave` leaves the stateful proof mode, cleaning up the proof state).
-```lean
-theorem mySum_correct_short (l : Array Nat) : mySum l = l.sum := by
-  generalize h : mySum l = x
-  apply Id.of_wp_run_eq h
-  mvcgen
-  case inv1 => exact ⇓⟨xs, out⟩ => ⌜xs.prefix.sum = out⌝
-  all_goals mleave; grind
+::::leanSection
+```lean -show
+variable {P : Prop} {σ : List (Type u)}
 ```
-This pattern is so common that {tactic}`mvcgen` comes with special syntax for it:
-```lean
-theorem mySum_correct_shorter (l : Array Nat) : mySum l = l.sum := by
-  generalize h : mySum l = x
-  apply Id.of_wp_run_eq h
-  mvcgen
-  invariants
-  · ⇓⟨xs, out⟩ => ⌜xs.prefix.sum = out⌝
-  with grind
+Ordinary propositions that do not mention the state can be used as stateful predicates by adding a trivial universal quantification.
+This is written with the syntax {lean (type := "SPred σ")}`⌜P⌝`, which is syntactic sugar for {name}`SPred.pure`.
+:::syntax term (title := "Notation for `SPred`") (namespace := Std.Do)
+```grammar
+⌜$_:term⌝
 ```
-The {keyword}`mvcgen invariants `{lit}`...`{keyword}` with `{lit}`...` is an abbreviation for the
-tactic sequence {keyword}`mvcgen; case`{lit}` inv1 => ...`{keyword}`; all_goals mleave; grind`
-above. It is the form that we will be using from now on.
+{includeDocstring Std.Do.«term⌜_⌝»}
 :::
+::::
+
+{docstring SPred.pure}
+
+:::example "Stateful Predicates"
+```imports -show
+import Std.Do
+import Std.Tactic.Do
+```
+```lean -show
+open Std.Do
 
-:::paragraph
-It is helpful to compare the proof of {name}`mySum_correct_shorter` to a traditional correctness proof:
+set_option mvcgen.warning false
 
+```
+The predicate {name}`ItIsSecret` expresses that a state of type {name}`String` is {lean}`"secret"`:
 ```lean
-theorem mySum_correct_vanilla (l : Array Nat) : mySum l = l.sum := by
-  -- Turn the array into a list
-  cases l with | mk l =>
-  -- Unfold `mySum` and rewrite `forIn` to `foldl`
-  simp [mySum]
-  -- Generalize the inductive hypothesis
-  suffices h : ∀ out, List.foldl (· + ·) out l = out + l.sum by simp [h]
-  -- Grind away
-  induction l with grind
+def ItIsSecret : SPred [String] := (⟨· = "secret"⟩)
```

Comment:
I suppose `⟨⟩` to inject into `ULift` works, but maybe it would be better to use corner brackets here so that we use `SPred.pure` instead of the defeq `SPred [] = ULift Prop`? Haven't tested whether that'll work with cdot though. In case it doesn't work, I'm happy with what you have.